### DISCUSSION TITLE
Failed to create casync bundle

### DISCUSSION THREAD

**Discussion by Sourabh1107** [rauc/rauc#892] [Q&A] [answered]
Hi,

Advice on how to work with rauc casync support would be helpful.

1. While converting rauc bundle to casync bundle, I am getting below error:

```
rauc-Message: 10:06:46.773: Verifying bundle payload...
rauc-Message: 10:06:51.986: Converting rootfs.ext4 to blob idx rootfs.ext4.caibx
Failed to create bundle: Failed to start casync: Failed to execute child process ?casync? (No such file or directory)
```

Is this complaining that casync doesn't exist?

Then I tried to build casync from [https://github.com/systemd/casync#building-casync](url) But that also results in error like:

```
The Meson build system
Version: 0.61.1
Source dir: /root/raspcm4/sources/casync-tmp/casync
Build dir: /root/raspcm4/sources/casync-tmp/casync/build
Build type: native build
Project name: casync
Project version: 2
qemu-aarch64: Could not open '/lib/ld-linux-aarch64.so.1': No such file or directory

meson.build:3:0: ERROR: Executables created by c compiler aarch64-linux-gcc are not runnable.
A full log can be found at /root/raspcm4/sources/casync-tmp/casync/build/meson-logs/meson-log.txt
```

Can you please let me know what is the proper way to use casync rauc support?

2. In the [https://rauc.readthedocs.io/en/latest/advanced.html#installing-casync-bundles](url) it's mentioned we need to have "remote casync chunk store". 
`rauc install https://server.example.com/deploy/bundle-20180112.raucb`

What is this server? Do we need to create our own server to store chunk files? Any example on how to create such casync chunk store would be helpful.

Can someone please clarify these doubts?

Your help will be much appreciated.

Thanks in advance

P.S: I am using Buildroot based Linux system to integrate RAUC with RPi CM4

**Comment by ejoerns** [comment]
> Hi,
> 
> Advice on how to work with rauc casync support would be helpful.
> 
> 1. While converting rauc bundle to casync bundle, I am getting below error:
> 
> ```
> rauc-Message: 10:06:46.773: Verifying bundle payload...
> rauc-Message: 10:06:51.986: Converting rootfs.ext4 to blob idx rootfs.ext4.caibx
> Failed to create bundle: Failed to start casync: Failed to execute child process ?casync? (No such file or directory)
> ```
> 
> Is this complaining that casync doesn't exist?

Exactly, yes.

> Then I tried to build casync from [https://github.com/systemd/casync#building-casync](url) But that also results in error like:
> 
> ```
> The Meson build system
> Version: 0.61.1
> Source dir: /root/raspcm4/sources/casync-tmp/casync
> Build dir: /root/raspcm4/sources/casync-tmp/casync/build
> Build type: native build
> Project name: casync
> Project version: 2
> qemu-aarch64: Could not open '/lib/ld-linux-aarch64.so.1': No such file or directory
> 
> meson.build:3:0: ERROR: Executables created by c compiler aarch64-linux-gcc are not runnable.
> A full log can be found at /root/raspcm4/sources/casync-tmp/casync/build/meson-logs/meson-log.txt
> ```
> 
> Can you please let me know what is the proper way to use casync rauc support?

Did you create a Buildroot recipe for building casync? I guess Buildroot should be able to properly set up meson for cross-compilation.

There is also a ready-to-use recipe in buildroot already I see:

https://github.com/buildroot/buildroot/tree/master/package/casync

> 2. In the [https://rauc.readthedocs.io/en/latest/advanced.html#installing-casync-bundles](url) it's mentioned we need to have "remote casync chunk store".
>    `rauc install https://server.example.com/deploy/bundle-20180112.raucb`
> 
> What is this server? Do we need to create our own server to store chunk files? Any example on how to create such casync chunk store would be helpful.
>

Did you have a look at https://rauc.readthedocs.io/en/latest/advanced.html#rauc-casync-support ?

I would highly recommend reading this first.

The server is the server where you download casync bundles and chunks from. No special requirements.
I would assume that you have a server anyway in you setup / use case when you think about using casync?

> Can someone please clarify these doubts?
> 
> Your help will be much appreciated.
> 
> Thanks in advance
> 
> P.S: I am using Buildroot based Linux system to integrate RAUC with RPi CM4

**Comment by Sourabh1107** [comment]
Hello @ejoerns thanks for the quick response.

> Did you create a Buildroot recipe for building casync? I guess Buildroot should be able to properly set up meson for cross-compilation.
> 
> There is also a ready-to-use recipe in buildroot already I see:
> 
> https://github.com/buildroot/buildroot/tree/master/package/casync

Yes, I have selected `BR2_PACKAGE_CASYNC` and `BR2_PACKAGE_HOST_MESON_TOOLS` in `menuconfig` But now this leads to some other error like:

`Found pkg-config: /root/raspcm4/sources/output/host/bin/pkg-config (1.6.3)`
`Run-time dependency liblzma found: YES 5.2.5`
`Run-time dependency zlib found: YES 1.2.11`
`Run-time dependency libzstd found: YES 1.5.2`
`Run-time dependency libcurl found: YES 7.81.0`
`Run-time dependency openssl found: YES 1.1.1m`

`meson.build:135:0: ERROR: C shared or static library 'acl' not found`

But `BR2_PACKAGE_ACL=y` is already selected.

Can you please let me know what is the proper package that should be selected in Buildroot?

> Did you have a look at https://rauc.readthedocs.io/en/latest/advanced.html#rauc-casync-support ?
> 
> I would highly recommend reading this first.
> 
> The server is the server where you download casync bundles and chunks from. No special requirements.
> I would assume that you have a server anyway in you setup / use case when you think about using casync?

Yes, I am following the documentation. Okay I do have a hawkBit server, can I use the same for casync? If not, what kind of server is needed for casync? Any example would be helpful.

Thanks in advance.

**Reply by ejoerns** [reply]
> Hello @ejoerns thanks for the quick response.
> 
> > Did you create a Buildroot recipe for building casync? I guess Buildroot should be able to properly set up meson for cross-compilation.
> > There is also a ready-to-use recipe in buildroot already I see:
> > https://github.com/buildroot/buildroot/tree/master/package/casync
> 
> Yes, I have selected `BR2_PACKAGE_CASYNC` and `BR2_PACKAGE_HOST_MESON_TOOLS` in `menuconfig` But now this leads to some other error like:
> 
> `Found pkg-config: /root/raspcm4/sources/output/host/bin/pkg-config (1.6.3)` `Run-time dependency liblzma found: YES 5.2.5` `Run-time dependency zlib found: YES 1.2.11` `Run-time dependency libzstd found: YES 1.5.2` `Run-time dependency libcurl found: YES 7.81.0` `Run-time dependency openssl found: YES 1.1.1m`
> 
> `meson.build:135:0: ERROR: C shared or static library 'acl' not found`
> 
> But `BR2_PACKAGE_ACL=y` is already selected.
> 
> Can you please let me know what is the proper package that should be selected in Buildroot?

I'm all but a buildroot expert. All I can see is that there is a line in the recipe mentioning

```
CASYNC_DEPENDENCIES = acl libcurl openssl
```

thus I would expect the dependency to be present. Maybe its just a lookup failure in meson?

But, as already suggested in https://github.com/systemd/casync/issues/256, asking in a buildroot channels sounds like the proper way to go..

> > Did you have a look at https://rauc.readthedocs.io/en/latest/advanced.html#rauc-casync-support ?
> > I would highly recommend reading this first.
> > The server is the server where you download casync bundles and chunks from. No special requirements.
> > I would assume that you have a server anyway in you setup / use case when you think about using casync?
> 
> Yes, I am following the documentation. Okay I do have a hawkBit server, can I use the same for casync? If not, what kind of server is needed for casync? Any example would be helpful.
> 
> Thanks in advance.

Server means something you can talk HTTP(S) to and that allows you to download files. No matter if you take an Apache, nginx, lighttpd or something else.

Casync tries to download a chunk at an URL that is determined by the base store URL and the chunk hash.
If your server can be reached under ``http://example.com/store``, and casync attempts to fetch the chunk ``4bda8badd9d791e13f5da38120af6c428ed92d44ffe19b031d46d6556fbcf970``, it will make a GET call on the URL ``http://example.com/store/4bda/4bda8badd9d791e13f5da38120af6c428ed92d44ffe19b031d46d6556fbcf970.cacnk``.

You an simpy let your chunk servers root path point to the casync `.castr` directory created by `rauc convert`.

hawkBit instead is a full deployment server that manages devices, provides a client REST API, etc.
Using hawkBit with casync should be possible, but is more or less decoupled. If you use hawkbit, then the bundle will be downloaded via the URL provided by hawkbit. The chunk store however then needs to be provided independently from hawkBit.

**Comment by Sourabh1107** [comment]
Hello @ejoerns thanks for the update.

**Update**

As of now Buildroot will create "casync" as target-only package which means casync can only be part of the image. It won't create a host package.

[https://github.com/buildroot/buildroot/tree/master/package/casync](url) This will create target package.

**Reply by ejoerns** [reply]
@Sourabh1107 Oh, thanks for clarifying. I didn't notice indeed.

Thus you can build casync for the target and fail to manuall build it for the host only, right?

If you build it outside Buildroot, then the standard way should work indeed (or needs to be fixed).

**Comment by Sourabh1107** [comment]
> Thus you can build casync for the target and fail to manually build it for the host only, right?

Yes.  I made some progress but I am facing errors like:

`ninja: Entering directory build'`
`[1/23] Linking target casync-http`
`FAILED: casync-http`
`cc  -o casync-http casync-http.p/src_casync-http.c.o -Wl,--as-needed -Wl,--no-undefined -Wl,`-`rpath,/root/raspcm4/sources/output/host/bin/../aarch64-buildroot-linux-gnu/sysroot/usr/lib -Wl,-rpath`-`link,/root/raspcm4/sources/output/host/bin/../aarch64-buildroot-linux-gnu/sysroot/usr/lib -Wl,--start-group src/libshared.a` `/root/raspcm4/sources/output/host/bin/../aarch64-buildroot-linux-gnu/sysroot/usr/lib/libcurl.so` `/root/raspcm4/sources/output/host/bin/../aarch64-buildroot-linux-gnu/sysroot/usr/lib/libssl.so` `/root/raspcm4/sources/output/host/bin/../aarch64-buildroot-linux-gnu/sysroot/usr/lib/libcrypto.so` `/root/raspcm4/sources/output/host/bin/../aarch64-buildroot-linux-gnu/sysroot/usr/lib/liblzma.so` `/root/raspcm4/sources/output/host/bin/../aarch64-buildroot-linux-gnu/sysroot/usr/lib/libz.so` `/root/raspcm4/sources/output/host/bin/../aarch64-buildroot-linux-gnu/sysroot/usr/lib/libzstd.so -lm -Wl,--end-group`
`/usr/bin/ld: /root/raspcm4/sources/output/host/bin/../aarch64-buildroot-linux-gnu/sysroot/usr/lib/libcurl.so: error adding symbols: file in wrong format`
`collect2: error: ld returned 1 exit status`

Any idea on how to proceed further will be helpful.

**Comment by Sourabh1107** [comment, marked as answer]
**Update**

In order to build casync for host, we need to modify buildroot/package/casync/casync.mk file to include `HOST_CASYNC_DEPENDENCIES = host-acl host-libcurl host-openssl` below `CASYNC_DEPENDENCIES = acl libcurl openssl` and `$(eval $(host-meson-package))` below `$(eval $(meson-package))`. Then you need to add `BR2_PACKAGE_HOST_CASYNC=YES` to the config and the related dependencies.

Then we can run `make host-casync` to build for host-casync.

But this will lead to further errors like: ` make: * No rule to make target 'host-libfuse'`

Again the issue is "fuse" is not available as host package in Buildroot. So, it looks like building a host package in Buildroot is not so straightforward.

So, building casync outside Buildroot env much simpler solution.

For Ubuntu, there is already "casync package" available. Installing such package is enough to convert rauc bundles to casync compatible bundles.

**Comment by Sourabh1107** [comment]
Hello @ejoerns Is it also possible to place chunk index files to Amazon s3 bucket instead of web server like lighttpd?

Any idea if we could achieve this would be insightful.

**Reply by ejoerns** [reply]
Maybe you would like to have a look at https://github.com/folbricht/desync which explicitly lists S3 as storage option.

**Comment by Sourabh1107** [comment]
Hello @ejoerns I am able to use casync with S3 bucket and rauc can install bundles from S3. 

> hawkBit instead is a full deployment server that manages devices, provides a client REST API, etc.
> Using hawkBit with casync should be possible, but is more or less decoupled. If you use hawkbit, then the bundle will be downloaded via the URL provided by hawkbit. The chunk store however then needs to be provided independently from hawkBit.

As of now, chunk files(.castr) are in S3 along with converted rauc bundle. But we have a hawkbit which is used for normal rauc bundles. 
But is it possible to have only chunk files in S3 and use hawkbit to install casync bundle? How will rauc knows about chunk store if we make it independent?

Thanks in advance.
